tests/run/i10511.scala

```diff
@@ -0,0 +1,31 @@
+enum Bool {
+  case True
+  case False
+
+  // just to make sure we are using reference equality
+  override def equals(a: Any) = false
+
+}
+
+import Bool._
+
+type Not[B <: Bool] = B match {
+  case True.type => False.type
+  case False.type => True.type
+}
+
+def not[B <: Bool & Singleton](b: B): Not[B] = b match {
```

Comment:
It's nice that `Singleton` helps inference here, the following compiles as expected:

```scala
not(True): False.type
```

Although it's a bit fragile:

```scala
val vrais = True
not(vrais): False.type // error, we lost the singleton type...
```

Reply:
yeah its true with enums you must do a lot of work to avoid the widening but thats something hopefully we can tune progressively without ruining bincompat